p2p/ipld/read.go

```diff
@@ -3,16 +3,215 @@ package ipld
 import (
 	"context"
 	"errors"
+	"fmt"
 	"math"
 
 	"github.com/ipfs/go-cid"
 	coreiface "github.com/ipfs/interface-go-ipfs-core"
 	"github.com/ipfs/interface-go-ipfs-core/path"
+	"github.com/lazyledger/lazyledger-core/p2p/ipld/plugin/nodes"
+	"github.com/lazyledger/lazyledger-core/types"
+	"github.com/lazyledger/rsmt2d"
 )
 
-// /////////////////////////////////////
+// //////////////////////////////////////
+//	Retrieve Block Data
+// ////////////////////////////////////
+
+const baseErrorMsg = "failure to retrieve block data:"
+
+var ErrEncounteredTooManyErrors = fmt.Errorf("%s %s", baseErrorMsg, "encountered too many errors")
+var ErrTimeout = fmt.Errorf("%s %s", baseErrorMsg, "timeout")
+
+// RetrieveBlockData asynchronously fetches block data until the underlying extended
+// data square can be restored and returned.
+func RetrieveBlockData(
+	ctx context.Context,
+	dah *types.DataAvailabilityHeader,
+	api coreiface.CoreAPI,
+	codec rsmt2d.CodecType,
+) (types.Data, error) {
+	edsWidth := uint32(len(dah.ColumnRoots))
+	originalSquareWidth := edsWidth / 2
+
+	// keep track of leaves using thread safe counter
+	sc := newshareCounter(ctx, edsWidth)
+
+	// convert the row and col roots into Cids
+	rowRoots := dah.RowsRoots.Bytes()
+	rowCids, err := rootsToCids(rowRoots)
+	if err != nil {
+		return types.Data{}, err
+	}
+
+	colRoots := dah.ColumnRoots.Bytes()
+	colCids, err := rootsToCids(colRoots)
+	if err != nil {
+		return types.Data{}, err
+	}
+
+	// async attempt to fetch each leaf
+	for axisIdx := uint32(0); axisIdx < uint32(len(rowCids)); axisIdx++ {
+		for idx := uint32(0); idx < edsWidth; idx++ {
+			go sc.retrieveShare(colCids[axisIdx], false, axisIdx, idx, api)
+			go sc.retrieveShare(rowCids[axisIdx], true, axisIdx, idx, api)
+		}
+	}
+
+	// wait until enough data has been collected, too many errors encountered,
+	// or the timeout is reached
+	err = sc.wait()
```

Comment:
So, OK, this wait() is supposed to return when enough data was requested. That makes sense. But this doesn't mean that the many (8*origSquareWidth*origSquareWidth) above started go-routines are canceled immediately after that, or at least I don't see how that works exactly. The first operation in `go retrieveShare` is `GetLeafData` which will block until the data was actually read from the network. So the many go-routines continue to run but what we'd want is to cancel all these go-routines asap when we have enough data.

It's probably better to only start as many as needed instead.